RecoLocalFastTime/FTLRecProducers/plugins/MTDTrackingRecHitProducer.cc

```diff
@@ -0,0 +1,154 @@
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/FTLRecHit/interface/FTLClusterCollections.h"
+
+#include "Geometry/Records/interface/MTDDigiGeometryRecord.h"
+#include "Geometry/CommonTopologies/interface/Topology.h"
+#include "Geometry/MTDGeometryBuilder/interface/MTDGeometry.h"
+#include "DataFormats/TrackerRecHit2D/interface/MTDTrackingRecHit.h"
+#include "DataFormats/GeometryCommonDetAlgo/interface/MeasurementError.h"
+#include "DataFormats/GeometryCommonDetAlgo/interface/MeasurementPoint.h"
+
+#include "RecoLocalFastTime/Records/interface/MTDCPERecord.h"
+#include "RecoLocalFastTime/FTLClusterizer/interface/MTDClusterParameterEstimator.h"
+#include "RecoLocalFastTime/FTLClusterizer/interface/MTDCPEBase.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+//#define DEBUG_ENABLED 
+#ifdef DEBUG_ENABLED
+#define DEBUG(x) do { std::cout << x << std::endl; } while (0)
+#else
+#define DEBUG(x)
+#endif
+
+class MTDTrackingRecHitProducer : public edm::stream::EDProducer<> {
+  
+ public:
+  explicit MTDTrackingRecHitProducer(const edm::ParameterSet& ps);
+  ~MTDTrackingRecHitProducer() override;
+  static void fillDescriptions(edm::ConfigurationDescriptions & descriptions);
+
+  void produce(edm::Event& evt, const edm::EventSetup& es) override;
+  void run(edm::Handle<edmNew::DetSetVector<FTLCluster> >  inputHandle,
+	   MTDTrackingDetSetVector &output);
+  
+ private:
+  const edm::EDGetTokenT<FTLClusterCollection> ftlbClusters_; // collection of barrel digis
+  const edm::EDGetTokenT<FTLClusterCollection> ftleClusters_; // collection of endcap digis
+    
+  const MTDGeometry* geom_;
+  const MTDClusterParameterEstimator* cpe_;
+};
+
+MTDTrackingRecHitProducer::MTDTrackingRecHitProducer(const edm::ParameterSet& ps) :
+  ftlbClusters_( consumes<FTLClusterCollection>( ps.getParameter<edm::InputTag>("barrelClusters") ) ),
+  ftleClusters_( consumes<FTLClusterCollection>( ps.getParameter<edm::InputTag>("endcapClusters") ) )
+{  
+  produces< MTDTrackingDetSetVector >();
+}
+
+// Configuration descriptions
+void
+MTDTrackingRecHitProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("barrelClusters", edm::InputTag("mtdClusters:FTLBarrel"));
+  desc.add<edm::InputTag>("endcapClusters", edm::InputTag("mtdClusters:FTLEndcap"));
+  descriptions.add("mtdTrackingRecHitProducer", desc);
+}
+
+MTDTrackingRecHitProducer::~MTDTrackingRecHitProducer() {
+}
+
+void
+MTDTrackingRecHitProducer::produce(edm::Event& evt, const edm::EventSetup& es) {
+  
+  edm::ESHandle<MTDGeometry> geom;
+  es.get<MTDDigiGeometryRecord>().get(geom);
+  geom_ = geom.product();
+
+  edm::ESHandle<MTDClusterParameterEstimator> cpe;
+  es.get<MTDCPERecord>().get("MTDCPEBase",cpe);
+  cpe_ = cpe.product();
+  
+  if ( ! geom_ ) 
+    {
+      throw cms::Exception("MTDTrackingRecHitProducer") << "Geometry is not available -- can't run!";
```

Comment:
`ESHandle<T>::product()` will already throw if the product is not valid, so this check and exception is redundant.